### DISCUSSION TITLE
Replace permission filter rules with custom code for complex permissions. New hook?

### DISCUSSION THREAD

**Discussion by wyteco** [directus/directus#11550] [Feature Requests]
Hey!

I posted a question over on Discord but noticed that the context needed might be worthy of its own GitHub discussion. So here I am 👋

# Our Setup
We have some collections whose items owned by a company can be shared by that company with other users.

So each `resource_*` collection has some kind of ownership field (`company_owner`) which is a M2O relation to `companies`.

`companies` can store `access_keys` which is a collection coming from a `companies` M2A relation to `resource_*`.

This `access_keys` (junction) collection has some additional fields:
    - M2O to `directus_users` (the user whom the key has been issued to)
    - `type` a string with the action  ("read", "write", etc.)

# Challenge
To check whether the user can access a resource I need to check whether the owner company of that resource issued an `access_keys` item which includes the user as well as the correct `type`.

**E.g. the user wants to read a `resource_trainings` item:**
For the user to be able to read the training, the owner company of that training needs to have an access key created linked to that training and to that user with `type = "read"`

# Problem & What I tried
My first approach using the filter rules:

`<resource_*>` > Access Keys > User > ID equals `$CURRENT_USER.id`

For this to work I'd need to store the access keys in the associated `<resource_*>` collections as well (I don't know whether that is possible in M2A relations)
Also I'd need to check that **specific** access key's `type` value to equal the requested action. Which is **AFAIK not possible with filter rules** as I can only check for _any_ access key's `type` value.

## Option 1 (using filter hooks)
I would have to force the user to include the `company_owner` field of the resource so I can access it in the `payload` and make a 2nd run to the database to get the company's access keys and check them against the requesting user, the requested resource and the action.
This could work but might also be kind of slow because of the multiple database runs.
There would be logic for every `<resource_*>.items.*` filter.

## Option 2 (custom endpoints)
I would need to (1) disable the default endpoints and (2) write some custom logic to re-implement the directus endpoint behaviour just with my permission-checking in between.
For task [1] I could only imagine to set all permissions to `No Access` for the user role.
Would that also block `ItemsService` which I would need to use in my custom logic? (user's `accountability`)

---

I am really stuck on this problem and would greatly appreciate any ideas on how to solve this.
Neither filter hooks nor custom endpoints sound like the best solutions to me.

Implementing a new hook to replace/ bypass the filter rules and execute own logic however might be the way to go. I thought about forking directus and implementing a custom solution for us but wanted to discuss that topic with you first as I am not familiar with the codebase and think that forking for a never-to-be-merged custom solution is kind of useless. Especially if you have any ideas on how to make this useful for more directus users.

Thank you

**Comment by rijkvanzanten** [comment]
Heya! Thanks for your detailed insight, really interesting use case you got there. Took me a couple reads to fully grasp it ^^

Just to expand on the two missing things you already pointed out: In order for this to work out of the box, the following two things would need to be added:

1) the "reverse" relationship for the many-to-any field type (like you pointed out in "_I'd need to store the access keys_"). With that, I think you would've been able to configure it by following the relationship back from the original resources table:

```
_and:
  access_tokens (<- field type that doesn't currently exist)
    user
      _eq $CURRENT_USER 
```

and;

2) Nested _and filter queries under a relational field (to allow for the any vs specific item filter).

When using the top level and, the following:

```
_and:
  access_tokens (<- field type that doesn't currently exist)
    user 
      _eq $CURRENT_USER 
  access_tokens (<- field type that doesn't currently exist)
    type 
      _eq 'read'
```

would mean: "where any of the associated access tokens is for user $CURRENT_USER and where any of the associated access tokens is of type 'read'". Instead, you're looking for "where any of the access tokens is for user $CURRENT_USER and is of type 'read'", which would look like follows:

```
access_tokens
  _and
    user 
      _eq $CURRENT_USER 
    type 
      _eq 'read'
```

---

Now for the question: How to solve for this anyways? My first answer would be: Lets implement those two missing features, and keep using the core systems 😄 That's a win win for everybody involved. That being said, these two features aren't _that_ trivial to add, so it would be a little much to say _Just open a PR!_. 

Between the two options you propose, I think only (1) is realistic. If you were to end up forking Directus and replace every endpoint with your own implementation, you'll most likely end up breaking the app (as it's built for the current system ones) and will probably run into a lot of trouble keeping your fork up to date. 

Between the hook, endpoints, and implementing those two missing features, I think the hook is probably the simplest. That being said, you're correct in pointing out the performance penalty in having to check for that extra layer of permissions after Directus' own permissions system already kicked in.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
